Smash apart the workspace into more focused crates

thomaseizinger opened the pull request:
TODO:

- [x] Ensure each commit builds, updates lockfile correctly and is formatted
- [x] Extract test suite into dedicated crate
- [x] Extract a `tokio-tasks` crate for the `Tasks` and `tokio_ext` stuff
- [x] Extract `xtras` crate
- [x] Write up document explaining crate structure and incentives?

thomaseizinger committed:
Configure global lints in .cargo/config

In the absence of a way to configure lints across a workspace, we use
a technique described in https://github.com/EmbarkStudios/rust-ecosystem/pull/68.

This allows us to move the lints that should unconditionally apply
to all crates to this one config, unblocking the split up of the mega
crate into multiple smaller ones.

thomaseizinger committed:
Extract `rocket_basicauth` crate

Turns out the `auth` module was still hardwired to `itchysats` username.
As part of the move, make the username configurable.

thomaseizinger committed:
Decouple `model` module from rocket

If we want to untangle our dependencies, we can't depend on stuff
like rocket in our model code.

thomaseizinger committed:
Don't import `time` from rocket re-export

thomaseizinger committed:
Extract rust-embed-rocket crate

The code in there is shared between both binaries. If we want to
extract a dedicated crate per binary, this code needs to go into
a dedicated crate.

thomaseizinger committed:
Swap `derive_more` for `parse_display`

This one support deriving `FromStr` as well.

thomaseizinger committed:
Don't implement `FromParam` for `CfdAction`

If we want to decouple the rocket dependency from our core crate, we
cannot have this implementation. This looks like a step backwards but
actually, we should solve this by not parsing these actions at all and
instead send links to the clients.

See https://github.com/itchysats/itchysats/issues/955.

thomaseizinger committed:
Don't print version in test

We always know which version a test is running at. Locally, you can
check with `git status` and each CI run is also pinned to a commit hash.

Removing this line allows us to split apart the daemon crate into
multiple and tear apart the dependencies.

thomaseizinger committed:
Extract `taker` and `maker` crates

To actually achieve this, we need to extract another crate
`daemon-shared`. This is not ideal but in the long run, we can try to
get rid of some of this code in there through various means. Some of the
options are documented in the README of the crate.

thomaseizinger committed:
Move Heartbeat to taker

We only use this there so no need to have this in `daemon-shared`.

klochowicz commented on .cargo/config.toml:
```diff
@@ -4,3 +4,11 @@ linker = "aarch64-linux-gnu-gcc"
 dev-maker = "run --bin maker -- testnet"
 dev-taker = "run --bin taker -- --maker localhost:9999 --maker-id 10d4ba2ac3f7a22da4009d813ff1bc3f404dfe2cc93a32bedf1512aa9951c95e testnet" # Maker ID matches seed found in `testnet/maker_seed`
+
+# Inspired by https://github.com/EmbarkStudios/rust-ecosystem/pull/68.
```
yup, this seems to be the most common way people configure lints in workspaces...

klochowicz commented on daemon/Cargo.toml:
```diff
@@ -12,7 +12,6 @@ bytes = "1"
 chrono = { version = "0.4", features = ["serde"] }
 clap = { version = "3.0.10", features = ["derive"] }
 derivative = "2"
-derive_more = { version = "0.99.17", default-features = false, features = ["display"] }
```
curious about this change - `derive_more` seems to be way more popular and mature? and I can't see any different options from this one? (btw: there's *so many* boilerplate `Display` manual impl, we shouldn't really bother with them unless we're doing something funky....)

thomaseizinger replied:
We only used it for `Display` and `FromStr` and `derive_more` [doesn't have](https://jeltef.github.io/derive_more/derive_more/from_str.html#enums) `FromStr` for enums ... :(

It is easily swapped if we want a different one!

klochowicz replied:
yeah, I know it's easily swapped, I was just curious as I suspected I missed something :) thank you for explanation!

klochowicz commented on daemon-shared/Cargo.toml:
```diff
@@ -0,0 +1,20 @@
+[package]
+name = "daemon-shared"
```
naming suggestions: "bin-shared", "app-shared"?  It would be more self-descriptive, as we're sharing the code among binaries, as you described below

thomaseizinger replied:
Yeah good call!

klochowicz reviewed (commented):
great start. excited to reap benefits of this 😎  have you noticed any build time wins so far?

thomaseizinger committed:
Extract tests into dedicated crate

This allows us to manage the harness code as a separate crate and
also further simplifies tearing apart the `daemon` crate into `maker`
and `taker`.

thomaseizinger committed:
Move wait_next_state into harness code

luckysori reviewed (approved):
I'm very happy with this direction.

klochowicz reviewed (approved):
I really appreciate the granularity of commits - they made the review a breeze! 💯

luckysori committed:
Split FutureExt into two traits

One half of the original trait is only relevant for `Tasks`, so it can
be private to daemon.

The other half is used in `daemon-tests`, but to reduce the API
footprint of `daemon`, we inline the functionality in the one place
where it is used in `daemon-tests`.

luckysori committed:
Extract tokio-tasks crate

This functionality is used across several crates and it's
self-contained. Doing this further reduces the API footprint of
`daemon`.

luckysori committed:
Extract xtras crate

These extensions to `xtra` are only used in `daemon` at the moment,
but they are definitely unrelated to the functionality offered by the
`daemon` crate.

luckysori committed:
Add supervisor::Actor to xtras

The `supervisor::Actor` is a pattern generally applicable to actor
frameworks, so it doesn't apper to belong in `daemon`.

Unfortunately, we need to introduce a dependency to `xtras` in
`maker`. Maybe the `MakerActorSystem` can be reworked to remove this
dependency, or perhaps the dependency is fine.

Fixes https://github.com/itchysats/itchysats/issues/952.

luckysori committed:
Add document explaining crate structure

thomaseizinger commented on maker/Cargo.toml:
```diff
@@ -22,6 +22,7 @@ tokio-tasks = { path = "../tokio-tasks" }
 tracing = { version = "0.1" }
 uuid = "0.8"
 xtra = { version = "0.6" }
+xtras = { path = "../xtras" }
```
I think this is fine for now, we already transitively depend on this.

thomaseizinger commented on CRATE-STRUCTURE.md:
```diff
@@ -0,0 +1,30 @@
+## Why so many crates?
```
Nice document :)

luckysori replied:
Haha thanks. Wasn't sure what to say, so feel free to add an modify!
